### PR TITLE
Fine tuning pipeline

### PR DESCRIPTION
**Pull request by malodetz**
Pipeline for fine-tuning with cross-validation and results serialization.

### CODE REVIEW

File: configs/new_psiminer_config.json [hunk 1]
```diff
@@ -0,0 +1,44 @@
+{
```
Comment (egor-bogomolov) [inline configs/new_psiminer_config.json]: Let's give the file a more meaningful name

File: scripts/fine_tune.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+def load_typed_code2seq(
+    checkpoint_path: str, config: DictConfig, vocabulary: Vocabulary
+) -> Tuple[TypedCode2Seq, TypedPathContextDataModule]:
+    model = TypedCode2Seq.load_from_checkpoint(checkpoint_path=checkpoint_path)
+    data_module = TypedPathContextDataModule(config, vocabulary)
+    return model, data_module
+
+
+KNOWN_MODELS = {"code2seq": load_code2seq, "code2class": load_code2class, "typed-code2seq": load_typed_code2seq}
```
Comment (egor-bogomolov) [inline scripts/fine_tune.py]: Maybe we can move this part to a separate file?

File: scripts/fine_tune.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    checkpoint = torch.load(model_path, map_location=torch.device("cpu"))
+    config = checkpoint["hyper_parameters"]["config"]
+    vocabulary = checkpoint["hyper_parameters"]["vocabulary"]
+    config.data_folder = dataset_path
+
+    if config.name not in KNOWN_MODELS:
+        print(f"Unknown model {config.name}, try one of {' '.join(KNOWN_MODELS.keys())}")
+        return ""
+
+    model, data_module = KNOWN_MODELS[config.name](model_path, config, vocabulary)
```
Comment (egor-bogomolov) [inline scripts/fine_tune.py]: Calling dict element is a bit confusing. Let's add a function that handles this and keep some clean logic in this method.

File: scripts/fine_tune.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+    return checkpoint_callback.best_model_path
+
+
+def fine_tune(dataset_path: str, model_path: str, folds_number: int):
+    """Do k-fold cross-validation and compare quality metrics before and after fine-tuning"""
+
+    project_name = os.path.basename(os.path.normpath(dataset_path))
+    start_path = os.path.join(dataset_path, "java-med-psi-no-types")
+    dataset = open(os.path.join(start_path, "java-med-psi-no-types.test.c2s"), "r")
```
Comment (egor-bogomolov) [inline scripts/fine_tune.py]: These constants should be passed as arguments.

File: scripts/fine_tune.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    samples = dataset.readlines()
+    fold_size = len(samples) // folds_number
```
Comment (egor-bogomolov) [inline scripts/fine_tune.py:90]: The result rounds to a wrong number (floor but most likely you need ceil)

**Review by egor-bogomolov — APPROVED** [review]
Awesome job! Let's refactor the code a little bit and merge.

File: scripts/fine_tune.py (hunk 4 above)
Comment (malodetz) [reply scripts/fine_tune.py]: Moved to constants

File: scripts/fine_tune.py (hunk 5 above)
Comment (malodetz) [reply scripts/fine_tune.py:90]: It can cause out-of-bounds on  samples  list

**Pull request merged**